numpy/_core/src/multiarray/refcount.c

```diff
@@ -340,6 +347,13 @@ PyArray_XDECREF(PyArrayObject *mp)
         }
     }
     else { /* handles misaligned data too */
+        if (PyArray_NDIM(mp) > NPY_MAXDIMS_LEGACY_ITERS) {
+            PyErr_Format(PyExc_RuntimeError,
+                    "this function only supports up to 32 dimensions but "
+                    "the array has %d.", PyArray_NDIM(mp));
+            return -1;
+        }
```

Comment:
We don't us this internally, might remove the function, but not sure yet.  In principle, anyone running into it should catch the error and not leak anything (normal array cleanup doesn't have this problem).